### ISSUE TITLE
Data join update pattern appears to not work right

### ISSUE THREAD

**Issue by bobmonteverde** [d3/d3-selection#75]
After entering new elements, the update code immediately after does not work.  For example:

```
var elems = d3_selection.select('.wrap').selectAll('div').data([1,2,3])

elems.enter().append('div').text(0)

elems.text(function(d) { return d })
```

Will show 0's not 1, 2, 3.

I made a jsfiddle to show the problem, on the second call (when the element are not being entered, just updated) the correct numbers are shown.  https://jsfiddle.net/1z4p7734/1/

**Comment by bobmonteverde** [comment]
Is there a new pattern where enter + update is:

```
elems.enter().append('div')
  .merge(elems)
    .text(...)
```

Seems like a pretty big change that's not obvious in the docs.  If this is the new way, guess this is a non-issue.

**Comment by mbostock** [comment]
I haven’t yet had a chance to write release notes for 4.0 detailing what has changed, but it’s on my list of things to do before releasing D3 4.0. (Arguably it’s the main thing holding up the release at this point.) I originally was writing them as I go, but it led to quite a bit of wasted effort as I iterated through the design, so I decided to defer release notes until the release. My apologies for the confusion.
